plugin/trino-hive/src/main/java/io/trino/plugin/hive/fs/TransactionScopeCachingDirectoryLister.java

```diff
@@ -0,0 +1,230 @@
+            {
+                // force cache entry weight update in case next file is cached
+                checkState(hasNext());
+                return cachedValueHolder.getCachedFile(fileIndex++).orElseThrow();
+            }
+        };
+    }
+
+    @VisibleForTesting
+    boolean isCached(Path path)
+    {
+        FetchingValueHolder cached = cache.getIfPresent(path);
+        return cached != null && cached.isFullyCached();
+    }
+
+    private static boolean isLocationPresent(Storage storage)
+    {
+        // Some Hive table types (e.g.: views) do not have a storage location
+        return storage.getOptionalLocation().isPresent() && isNotEmpty(storage.getLocation());
+    }
+
+    private static class FetchingValueHolder
+    {
+        private final List<LocatedFileStatus> cachedFiles = synchronizedList(new ArrayList<>());
+        @GuardedBy("this")
+        @Nullable
+        private RemoteIterator<LocatedFileStatus> fileIterator;
+        @GuardedBy("this")
+        @Nullable
+        private IOException exception;
+
+        public FetchingValueHolder(RemoteIterator<LocatedFileStatus> fileIterator)
+        {
+            this.fileIterator = requireNonNull(fileIterator, "fileIterator is null");
+        }
+
+        public synchronized boolean isFullyCached()
+        {
+            return fileIterator == null && exception == null;
+        }
+
+        public int getCachedFilesSize()
+        {
+            return cachedFiles.size();
+        }
+
+        public RemoteIterator<LocatedFileStatus> getCachedFiles()
+        {
+            checkState(isFullyCached());
+            return new SimpleRemoteIterator(cachedFiles.iterator());
+        }
+
+        public Optional<LocatedFileStatus> getCachedFile(int index)
+                throws IOException
+        {
+            int filesSize = cachedFiles.size();
+            checkArgument(index >= 0 && index <= filesSize);
+
+            if (index < filesSize) {
+                return Optional.of(cachedFiles.get(index));
+            }
+
+            return fetchNextCachedFile(index);
+        }
+
+        private synchronized Optional<LocatedFileStatus> fetchNextCachedFile(int index)
+                throws IOException
+        {
+            if (index < cachedFiles.size()) {
+                // file was fetched concurrently
+                return Optional.of(cachedFiles.get(index));
+            }
+
+            if (exception != null) {
+                throw exception;
+            }
+
+            try {
+                if (fileIterator == null || !fileIterator.hasNext()) {
+                    // no more files
+                    fileIterator = null;
+                    return Optional.empty();
+                }
+
+                LocatedFileStatus fileStatus = fileIterator.next();
+                cachedFiles.add(fileStatus);
+                return Optional.of(fileStatus);
+            }
+            catch (IOException exception) {
```

Comment:
catch RTE too

Reply:
Did this in `hasNext` that invalidates cache entry